api/model/src/main/java/org/projectnessie/model/ContentKey.java

```diff
@@ -165,22 +166,35 @@ public static ContentKey of(@JsonProperty("elements") List<String> elements) {
   @Value.Check
   protected void validate() {
     List<String> elements = getElements();
+    int sum = 0;
     for (String e : elements) {
       if (e == null) {
         throw new IllegalArgumentException(
-            String.format("Content key '%s' must not contain a null element.", elements));
+            format("Content key '%s' must not contain a null element.", elements));
       }
-      if (e.isEmpty()) {
+      int l = e.length();
+      sum += l;
+      if (l == 0) {
         throw new IllegalArgumentException(
-            String.format("Content key '%s' must not contain an empty element.", elements));
+            format("Content key '%s' must not contain an empty element.", elements));
       }
-      if (e.chars().anyMatch(i -> i < FIRST_ALLOWED_KEY_CHAR)) {
-        throw new IllegalArgumentException(
-            String.format(
-                "Content key '%s' must not contain characters less than 0x%2h.",
-                elements, FIRST_ALLOWED_KEY_CHAR));
+      for (int i = 0; i < l; i++) {
+        char c = e.charAt(i);
+        if (c < FIRST_ALLOWED_KEY_CHAR) {
+          throw new IllegalArgumentException(
+              format(
+                  "Content key '%s' must not contain characters less than 0x%2h.",
+                  elements, FIRST_ALLOWED_KEY_CHAR));
+        }
       }
     }
+    if (sum > MAX_LENGTH) {
+      throw new IllegalStateException("Key too long, max allowed length: " + MAX_LENGTH);
+    }
+    if (elements.size() > MAX_ELEMENTS) {
```

Comment:
This if block could be moved to line 169: there is no point in iterating over the elements, if we know that there are too many elements.

Reply:
That would optimize for the bad case though.